File: Top Interview Questions/src/38.count-and-say.cpp
/*
 * @lc app=leetcode id=38 lang=cpp
 *
 * [38] Count and Say
 */
#include <string>
using namespace std;
// @lc code=start
class Solution {
 public:
  string countAndSay(int n) {
    string digits = "1";

    for (int i = 0; i < n - 1; i++) {
      digits = read_and_write(digits);
    }

    return digits;
  }

  string read_and_write(string digits) {
    string result = "";

    int count = 1;
    char number = digits.front();

    for (int i = 1; i < digits.size(); i++) {
      if (number == digits[i]) {
        count++;
      } else {
        result += (to_string(count) + number);
        count = 1;
        number = digits[i];
      }
    }
    result += (to_string(count) + number);

    return result;
  }
};
// @lc code=end

File: Top Interview Questions/src/38.count-and-say.java
/*
 * @lc app=leetcode id=38 lang=java
 *
 * [38] Count and Say
 */

// @lc code=start
class Solution {
  public String countAndSay(int n) {
    if (n == 1) {
      return "1";
    }

    String target = "1";
    
    for (int i = 1; i < n; i++) {
      StringBuilder result = new StringBuilder();

      int count = 1;
      char currentNumber = target.charAt(0);
      for (int j = 1; j < target.length(); j++) {
        
        if (currentNumber == target.charAt(j)) {
          count++;
        } else {
          result.append(count);
          result.append(currentNumber);
          count = 1;
          currentNumber = target.charAt(j);
        }
      }

      result.append(count);
      result.append(currentNumber);
      target = result.toString();
    }

    return target;
  }
}
// @lc code=end

